test/integration/scheduler/taint_test.go

```diff
@@ -787,18 +790,47 @@ func TestTaintBasedEvictions(t *testing.T) {
 			}
 
 			for i := 0; i < nodeCount; i++ {
-				// Stop the neededNode's heartbeat goroutine.
-				if neededNode.Name == fmt.Sprintf("node-%d", i) {
+				// If current Node is not <neededNode>, don't bother - let the original goroutine
+				// update its Ready status regularly.
+				if neededNode.Name != fmt.Sprintf("node-%d", i) {
+					continue
+				}
+
+				// Current Node is neededNode.
+				// Need to distinguish NodeReady/False and NodeReady/Unknown.
+				c, err := nodeReadyStatus(test.nodeConditions)
+				if err != nil {
+					t.Error(err)
+				}
+				// If we try to update the node with condition NotReady/False, i.e. expect a NotReady:NoExecute taint
+				// we need to keep sending the update event to keep it alive, rather than just sending once.
+				if c == v1.ConditionFalse {
+					close(heartbeatChans[i])
+					// replace with a new goroutine and channel.
+					heartbeatChans[i] = make(chan struct{})
+					go func(i int) {
+						for {
+							select {
+							case <-heartbeatChans[i]:
+								return
+							case <-time.Tick(heartbeatInternal):
+								nodeCopy := nodeCopyWithConditions(neededNode, test.nodeConditions)
+								if err := updateNodeStatus(cs, nodeCopy); err != nil {
+									t.Errorf("Cannot update node: %v", err)
+								}
+							}
+						}
+					}(i)
+				} else if c == v1.ConditionUnknown {
+					// If we try to update the node with condition NotReady/Unknown, i.e. expect a Unreachable:NoExecute taint,
+					// we need to only send the update event once to simulate the network unreachable scenario.
 					heartbeatChans[i] <- struct{}{}
```

Comment:
I might just be missing it, but do we still send a heartbeat to the other (notneeded) nodes? From what I can tell this loop [skips all the others](https://github.com/kubernetes/kubernetes/pull/84883/files#diff-d51d948272758524cc6289bc6f0b8daaR795-R797) so I can't find where the heartbeat is sent to those as well, though I could be wrong

Reply:
> do we still send a heartbeat to the other (notneeded) nodes?

Yes, in the beginning, when we instantiate the 3 goroutines, they're all sending NodeReady/True to APIServer: 


```go
			// Regularly send heartbeat event to APIServer so that the cluster doesn't enter fullyDisruption mode.
			// TODO(Huang-Wei): use "NodeDisruptionExclusion" feature to simply the below logic when it's beta.
			var heartbeatChans []chan struct{}
			for i := 0; i < nodeCount; i++ {
				heartbeatChans = append(heartbeatChans, make(chan struct{}))
			}
			for i := 0; i < nodeCount; i++ {
				// Spin up <nodeCount> goroutines to send heartbeat event to APIServer periodically.
				go func(i int) {
					for {
						select {
						case <-heartbeatChans[i]:
							return
						case <-time.Tick(heartbeatInternal):
							nodes[i].Status.Conditions = []v1.NodeCondition{
								{
									Type:              v1.NodeReady,
									Status:            v1.ConditionTrue,
									LastHeartbeatTime: metav1.Now(),
								},
							}
							updateNodeStatus(cs, nodes[i])
						}
					}
				}(i)
			}
```

And in this loop, if current node is "neededNode" we intentionally do (1) keep sending NodeReady/False events, or (2) only send NodeReady/Unknow once.

Reply:
sgtm, so the call to `updateNodeStatus` with now as the `lastHeartbeatTime` serves that purpose? I was looking for another spot where we actually passed something to the heartbeat channel, so I think I misunderstood that